### PR TITLE
SourceMgr: clean API for consistency and flexibility, faster SrcLoc conversion

### PR DESCRIPTION
**Pull request by chqrlie**
- add `string_buffer.Buf.detach()` to get buffer with ownership
- add `File.detach()` to get buffer with ownership
- simplify `string_buffer.Buf.newline()`
- add `File.openCount` and `File.is_const`
- unify `File` contents as `File.data_`
- improve CheckPoint system (simpler and much faster)
- remove `Max_open_files`
- remove `close_oldest()`

### CODE REVIEW

**Comment by bvdberg** [comment]
It looks a lot cleaner.. I found 2 issues:

- valgrind report a memleak, it looks like sm.index
- the refs dont work anymore. You can test this with:
 
`c2tags tools/c2loc.c2 142 5`

master branch:
found tools/c2loc.c2 120 22

this branch:
found tools/c2loc.c2 120 1

So the column is 1 instead of 22..
For some other locations it cannot find any result at all.
This also impacts the diagnostics, that dont show the whole line anymore, but only starting from the error location.

I marked the code issues in the sources files..

File: common/source_mgr.c2 [hunk 1]
```diff
@@ -40,105 +40,77 @@ public type Location struct {
...
-        return &f.checkpoints[f.checkpoint_count-1];
+fn void File.addCheckPoints(File* f) {
+    u32 ncheck = f.data_size / CheckPointSize;
+    CheckPoint* cp = stdlib.malloc((ncheck + 1) * sizeof(CheckPoint));
+    CheckPoint pos = { 0, 0 }
+    const char* src = f.data_;
+    f.checkpoints = cp;
+    while (ncheck-- > 0) {
+        *cp++ = pos;
+        pos = update_checkpoint(pos, src, CheckPointSize);
```
[the remark this answers is not in the document]
Comment (chqrlie) [reply common/source_mgr.c2:102]: My bad, the *optimized* version of `update_checkpoint()` was using the wrong logic.  We should have regression tests for the column numbers.

File: common/source_mgr.c2 [hunk 2]
```diff
@@ -175,46 +143,33 @@ public fn void SourceMgr.free(SourceMgr* sm) {
...
 }
 
-fn file_utils.Reader SourceMgr.openInternal(SourceMgr* sm, const char* filename, SrcLoc loc) {
+fn void* SourceMgr.openInternal(SourceMgr* sm, u32 name, SrcLoc loc, u32* psize) {
+    const char* filename = sm.pool.idx2str(name);
     file_utils.Reader file;
     if (file.open(filename)) {
-        sm.num_open++;
+        return file.detach(psize);
     } else {
```
[the remark this answers is not in the document]
Comment (chqrlie) [reply common/source_mgr.c2:173]: I was trying to minimize the diff with the current version, but I agree, let's make it lighter.

**Comment by bvdberg** [comment]
We can also do that in a separate patch.. i dont mind.
When I was using the non-optimized version (with the if stmt), the issue still seemed present..

File: common/source_mgr.c2 [hunk 3]
```diff
@@ -175,46 +143,33 @@ public fn void SourceMgr.free(SourceMgr* sm) {
...
+    sm.max_offset = sm.index_count * FileIndexSize;
 }
 
-fn file_utils.Reader SourceMgr.openInternal(SourceMgr* sm, const char* filename, SrcLoc loc) {
+fn void* SourceMgr.openInternal(SourceMgr* sm, u32 name, SrcLoc loc, u32* psize) {
+    const char* filename = sm.pool.idx2str(name);
     file_utils.Reader file;
     if (file.open(filename)) {
-        sm.num_open++;
+        return file.detach(psize);
```
[the remark this answers is not in the document]
Comment (chqrlie) [reply common/source_mgr.c2:172]: Or maybe the `file_utils` module should have a `loadFile()` function that returns the buffer, size and/or error.

Here is a proposal: 
* create a `file_utils.File` object that holds the pathname, a handle and the error code initialized from 1, 2 or 3 parts by corresponding init functions: `file_utils.File file.init(filename)`.
* this object would have a `load` and `store` functions to open, read and/or write the file contents directly and close the handle.
* possibly also an `open` and `close` functions to use as a regular file for `read` and `write` operations.
* `read` and `write` functions that use the system calls, handle the `EINTR` failure and partial writes gracefully.
* other functions for creating / deleting the file / directory, enumerate directory entries, get file attributes and length, etc.
* error handling would stay the same: the error code is stored in the object and can be retrieved as a code or an a explanatory string.
This is vaguely inspired by the java `File` class, but simpler.

**Comment by chqrlie** [comment]
> It looks a lot cleaner.. I found 2 issues:
> 
> * valgrind report a memleak, it looks like sm.index

Indeed, the old `index` was not freed upon reallocation.

> * the refs dont work anymore. You can test this with:
> 
> `c2tags tools/c2loc.c2 142 5`
> 
> master branch: found tools/c2loc.c2 120 22
> 
> this branch: found tools/c2loc.c2 120 1

Fixed.

> So the column is 1 instead of 22.. For some other locations it cannot find any result at all. This also impacts the diagnostics, that dont show the whole line anymore, but only starting from the error location.

Should be fixed too: since the columns were always `1`, some error messages would not show the line at all.

> I dont understand the openCount. It does not seem to be used? On File.close, it's just decremented. I would expect something like: if (--openCount == 0) file.clear(); Cant we remove that whole concept? Files are never closed, only SourceMgr.clear() 'closes' them.

The intent was to implement an unmapping scheme to limit the memory used. I shall remove this for now.

**Comment by chqrlie** [comment]
> We can also do the file_utils.Reader in a separate patch.. i dont mind.

Yes, we can discuss the `file_util` module separately.

> When I was using the non-optimized version (with the if stmt), the issue still seemed present..

Because you need to recompile twice :)  I always use `make rebuild` to avoid this pitfall.
